### PR TITLE
feat: Custom Applications multiple permissions

### PR DESCRIPTION
**Pull request by kark**
https://jira.commercetools.com/browse/SHIELD-580

In preparation for the Custom Applications Multiple Permissions support.

### CODE REVIEW

File: .changeset/cuddly-chefs-burn.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@commercetools-frontend/application-config': patch
+---
+
+Define new field `additionalOAuthScopes` in the Custom Application config.
```
Comment (CarlosCortizasCT) [inline .changeset/cuddly-chefs-burn.md:5]: Should we describe what this new field should be use for?
Or maybe add a link to the updated documentation?

File: packages/application-config/test/json-schema.spec.js [hunk 2]
```diff
@@ -152,6 +152,63 @@ describe('invalid configurations', () => {
       `"/oAuthScopes/manage/0 must match pattern \\"manage_(.*)\\""`
     );
   });
+  it('should validate that "additionalOAuthScopes" is an array', () => {
```
Comment (CarlosCortizasCT) [inline packages/application-config/test/json-schema.spec.js:155]: nit: Maybe another test to validate the additional oauth scopes object must contain the three required properties? (`name`, `view` and `manage`).

File: website/src/content/api-reference/application-config.mdx [hunk 3]
```diff
@@ -355,6 +355,78 @@ Using `manage_` OAuth Scopes always imply the corresponding `view_` OAuth Scope.
...
+    }
+  ]
+}
+```
+
+allows to send the additional permission information as claims scoped to each permission group:
+
+```
+view/movies:view_products view/merch:view_channels
+```
```
Comment (CarlosCortizasCT) [inline website/src/content/api-reference/application-config.mdx:393]: nit: I would omit this part since is talking about something more internal and may confuse our users.

File: website/src/content/api-reference/application-config.mdx [hunk 4]
```diff
@@ -355,6 +355,78 @@ Using `manage_` OAuth Scopes always imply the corresponding `view_` OAuth Scope.
 
 </Info>
 
+## `additionalOAuthScopes`
+
+The configuration for additional OAuth Scopes and user permissions extending [oAuthScopes configuration](#oauthscopes).
```
Comment (CarlosCortizasCT) [inline website/src/content/api-reference/application-config.mdx:360]: I'm missing some more context about what this field enables Custom Application developers to achieve.
Perhaps that's something to include in the [OAuth Scopes and user permissions](https://docs.commercetools.com/custom-applications/concepts/oauth-scopes-and-user-permissions) page.

Also, maybe it's worthy to add a new release note explaining this new feature.

**Review by CarlosCortizasCT — COMMENTED** [review]
This looks very promising 💪 

I just left some comments with some questions/suggestions.

File: packages/application-config/test/json-schema.spec.js (hunk 2 above)
Comment (kark) [reply packages/application-config/test/json-schema.spec.js:155]: Thanks! Considering that we already have a test verifying that `name` is provided, and having in mind that `view` and `manage` take `[]` as default, I guess we have that covered as I see it

File: website/src/content/api-reference/application-config.mdx (hunk 4 above)
Comment (kark) [reply website/src/content/api-reference/application-config.mdx:360]: Adding some explanation in the [OAuth Scopes and user permissions](https://docs.commercetools.com/custom-applications/concepts/oauth-scopes-and-user-permissions) page is a very good point 👍

File: packages/application-config/test/json-schema.spec.js (hunk 2 above)
Comment (ddouglasz) [reply packages/application-config/test/json-schema.spec.js:155]: Yes 👍, if `view` and `manage` have `[]` as default, with the test for `name` provided that I see, I guess that is covered.

File: website/src/content/api-reference/application-config.mdx [hunk 5]
```diff
@@ -355,6 +355,62 @@ Using `manage_` OAuth Scopes always imply the corresponding `view_` OAuth Scope.
...
+
+The configuration for [additional OAuth Scopes and user permissions](/concepts/oauth-scopes-and-user-permissions#additional-oauth-scopes) extending [oAuthScopes configuration](#oauthscopes).
+
+## `additionalOAuthScopes.name`
+
+A name used to build the permission keys based on the default permissions pair. 
+
+## `additionalOAuthScopes.view`
+
+A list of "view-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated to the `View` permission.
```
Comment (Rhotimee) [inline website/src/content/api-reference/application-config.mdx:368]:
```suggestion
A list of "view-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated with the additionalOAuthScopes `View` permission.
```

File: website/src/content/api-reference/application-config.mdx [hunk 6]
```diff
@@ -355,6 +355,62 @@ Using `manage_` OAuth Scopes always imply the corresponding `view_` OAuth Scope.
...
+      "view": ["view_channels"],
+      "manage": []
+    }
+  ]
+}
+```
+
+## `additionalOAuthScopes.manage`
+
+A list of "manage-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated with the `Manage` permission.
```
Comment (Rhotimee) [inline website/src/content/api-reference/application-config.mdx:389]:
```suggestion
A list of "manage-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated with the additionalOAuthScopes `Manage` permission.
```

File: website/src/content/concepts/oauth-scopes-and-user-permissions.mdx [hunk 7]
```diff
@@ -30,6 +30,37 @@ Notice here how the OAuth Scopes are grouped by the two fields `view` and `manag
 
 This grouping determines the **mapping and relation between OAuth Scopes and user permissions**.
 
+# Additional OAuth Scopes
+
+Defining `oAuthScopes` in the Custom Application config allows to use permissions limited to 1 unique pair (view/manage) specific to the Custom Application.
```
Comment (Rhotimee) [inline website/src/content/concepts/oauth-scopes-and-user-permissions.mdx:35]:
```suggestion
Defining `oAuthScopes` in the Custom Application config allows using permissions limited to 1 unique pair (view/manage) specific to the Custom Application.
```

File: website/src/content/concepts/oauth-scopes-and-user-permissions.mdx [hunk 8]
```diff
@@ -30,6 +30,37 @@ Notice here how the OAuth Scopes are grouped by the two fields `view` and `manag
...
+
+For more granular permissions, for example to allow team access to only certain parts or functionality of the Custom Application, [additional OAuth Scopes](https://docs.commercetools.com/api/scopes) can be requested as part of various permission groups.
```
Comment (Rhotimee) [inline website/src/content/concepts/oauth-scopes-and-user-permissions.mdx:37]:
```suggestion
For more granular permissions, for example, to allow the team access to only certain parts or functionality of the Custom Application, [additional OAuth Scopes](https://docs.commercetools.com/api/scopes) can be requested as part of various permission groups. 
```

**Review by Rhotimee — APPROVED** [review]
Looks good. Just some nitpicking.

File: .changeset/fresh-coins-shout.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (emmenko) [inline .changeset/fresh-coins-shout.md:1]: Maybe we should have 1 changeset in this case, as all the changes here relate to enabling a specific feature.

File: website/src/content/api-reference/application-config.mdx [hunk 10]
```diff
@@ -327,7 +327,7 @@ You can have "view-only" or "manage-only" OAuth Scopes and leave the other list
 
 ## `oAuthScopes.view`
 
-A list of "view-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated to the `View` permission.
+A list of "view-only" [OAuth Scopes](https://docs.commercetools.com/api/scopes) required by the Custom Application and associated with the `View` permission.
```
Comment (emmenko) [inline website/src/content/api-reference/application-config.mdx:330]: I would prefer the documentation changes to be defined in a separate PR. Main reason being that merging this PR will publish the docs changes but the packages are not being released yet.

So the docs changes can be merged separately once the packages have been released.

Can you extract these docs changes into a separate PR? Thanks 🙂

**Review by emmenko — COMMENTED** [review]
Have you had a chance to test the OIDC flow locally?

**Review by CarlosCortizasCT — APPROVED** [review]
Awesome job 🚀

**Review by emmenko — APPROVED** [review]
🚀
